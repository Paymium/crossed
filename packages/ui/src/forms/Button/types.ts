/**
 * Copyright (c) Paymium.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root of this projects source tree.
 */

import { CrossedMethods } from '@crossed/styled';
import { PressableProps } from 'react-native';

type Variants = 'primary' | 'tertiary' | 'secondary' | false;
export type ButtonProps = Omit<PressableProps, 'style'> & {
  /**
   * Select button variant
   * @default primary
   */
  variant?: Variants;
  /**
   * Select error colors
   * @type {boolean}
   */
  error?: boolean;
  /**
   * Loading state of button
   * @type {boolean}
   */
  loading?: boolean;
  /**
   * Extend style of button
   * @type {boolean}
   */
  style?: CrossedMethods<any>;
  /**
   * Disable size
   * @type {boolean}
   */
  size?: boolean;
};
