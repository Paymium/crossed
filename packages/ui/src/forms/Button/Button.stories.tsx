import type { Meta, StoryObj } from '@storybook/react';
import { View } from 'react-native';

import { Button } from './Button';

const ButtonMeta: Meta<typeof Button> = {
  title: 'Button',
  component: Button,
  render: (args) => (
    <Button {...args}>
      <Button.Text>Button</Button.Text>
    </Button>
  ),
  argTypes: {
    variant: {
      control: 'radio',
      options: ['primary', 'secondary', 'tertiary'],
    },
    error: { control: 'boolean' },
    loading: { control: 'boolean' },
  },
  args: {
    variant: 'primary',
  },
  decorators: [
    (Story) => (
      <View
        style={[
          { alignItems: 'center', justifyContent: 'center', flex: 1 },
          // { $$css: true, dark: 'light' },
        ]}
      >
        <Story />
      </View>
    ),
  ],
};

export default ButtonMeta;

export const Basic: StoryObj<typeof Button> = {
  args: {
    variant: 'primary',
    error: false,
    loading: false,
  },
};

export const Click: StoryObj<typeof Button> = {
  args: {
    variant: 'primary',
    error: false,
    loading: false,
  },
};
