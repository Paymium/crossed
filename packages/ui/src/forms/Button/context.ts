/**
 * Copyright (c) Paymium.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root of this projects source tree.
 */

import { createScope } from '@crossed/core';
import { Orientation, createCollection } from '@crossed/primitive';
import { createContext } from 'react';
import { ButtonProps } from './types';

export type ContextButton = {
  id: string;
};

export const [Provider, useContext] = createScope<ContextButton>(
  {} as ContextButton
);

export const buttonContext = createContext<
  Pick<ButtonProps, 'variant' | 'error' | 'disabled'> & {
    state?: {
      active?: boolean;
      hover?: boolean;
    };
  }
>({});

type ButtonGroupContext = {
  orientation: Orientation;
  grouped?: boolean;
};
export const [ProviderGroup, useContextGroup] = createScope<ButtonGroupContext>(
  { orientation: 'horizontal' } as ButtonGroupContext
);

const GROUP_NAME = 'ButtonGroup';

type ItemData = { id: string };
const [ButtonGroupCollection, useButtonGroupCollection] = createCollection<
  HTMLSpanElement,
  ItemData
>(GROUP_NAME);

export const {
  ItemSlot: ButtonGroupCollectionItemSlot,
  Provider: ButtonGroupCollectionProvider,
  Slot: ButtonGroupCollectionSlot,
} = ButtonGroupCollection;

export { ButtonGroupCollection, useButtonGroupCollection };
