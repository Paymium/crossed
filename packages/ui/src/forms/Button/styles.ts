/**
 * Copyright (c) Paymium.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root of this projects source tree.
 */

import { createStyles } from '@crossed/styled';

export const textStyles = createStyles((t) => ({
  default: {
    base: { fontWeight: 'bold' },
  },
  disabled: { base: { pointerEvents: 'none' } },
  primary: {
    'base': { color: t.components.Action.primary.default.text },
    ':hover': { color: t.components.Action.primary.hover.text },
    ':active': { color: t.components.Action.primary.active.text },
  },
  secondary: {
    'base': { color: t.components.Action.secondary.default.text },
    ':hover': { color: t.components.Action.secondary.hover.text },
    ':active': { color: t.components.Action.secondary.active.text },
  },
  tertiary: {
    'base': { color: t.components.Action.tertiary.default.text },
    ':hover': { color: t.components.Action.tertiary.hover.text },
    ':active': { color: t.components.Action.tertiary.active.text },
  },
  false: {},
}));

export const textDisabledStyles = createStyles((t) => ({
  primary: {
    base: { color: t.components.Action.primary.disabled.text },
  },
  secondary: {
    base: { color: t.components.Action.secondary.disabled.text },
  },
  tertiary: {
    base: { color: t.components.Action.tertiary.disabled.text, opacity: 0.5 },
  },
  false: {},
}));

export const textErrorStyles = createStyles((t) => ({
  primary: {
    base: { color: 'white' },
  },
  secondary: {
    'base': { color: t.colors.error.primary },
    ':hover': { color: t.colors.error.muted },
    ':active': { color: t.colors.error.satured },
    ':disabled': { color: t.colors.error.hight },
  },
  tertiary: {
    'base': { color: t.colors.error.primary },
    ':hover': { color: t.colors.error.muted },
    ':active': { color: t.colors.error.satured },
    ':disabled': { color: t.colors.error.hight },
  },
}));
export const textErrorDisabledStyles = createStyles((t) => ({
  primary: {
    base: { color: 'white' },
  },
  secondary: {
    base: { color: t.colors.error.hight },
  },
  tertiary: {
    base: { color: t.colors.error.hight },
  },
}));

export const buttonErrorStyles = createStyles(
  (t) =>
    ({
      error: {
        web: {
          ':focus-visible': {
            outlineColor: t.colors.error.primary,
          },
        },
      },
      primary: {
        'base': {
          backgroundColor: t.colors.error.primary,
          borderColor: t.colors.error.primary,
        },
        ':hover': {
          backgroundColor: t.colors.error.muted,
          borderColor: t.colors.error.muted,
        },
        ':active': {
          backgroundColor: t.colors.error.satured,
          borderColor: t.colors.error.satured,
        },
        ':disabled': {
          backgroundColor: t.colors.error.hight,
          borderColor: t.colors.error.hight,
        },
      },
      secondary: {
        'base': {
          borderColor: t.colors.error.primary,
          backgroundColor: t.colors.error.low,
        },
        ':hover': {
          borderColor: t.colors.error.muted,
          backgroundColor: t.colors.error.low,
        },
        ':active': {
          borderColor: t.colors.error.satured,
          backgroundColor: t.colors.error.hight,
        },
        ':disabled': { borderColor: t.colors.error.hight },
      },
      tertiary: { base: {} },
    }) as const
);

export const buttonSizeStyles = createStyles(() => ({
  false: { base: { height: 'auto' } },
  true: { base: { height: 44 } },
}));
export const buttonStyles = createStyles(
  (t) =>
    ({
      root: {
        base: {
          display: 'flex',
          paddingHorizontal: t.space.md,
          paddingVertical: t.space.sm,
          borderRadius: 7,
          borderWidth: 2,
          borderStyle: 'solid',
          borderColor: 'transparent',
          flexDirection: 'row',
          alignItems: 'center',
          justifyContent: 'center',
          backgroundColor: 'transparent',
          gap: 8,
        },
        web: {
          'base': {
            boxSizing: 'border-box',
          },
          ':focus-visible': {
            outlineWidth: '2px',
            outlineOffset: '2px',
            outlineStyle: 'solid',
            outlineColor: t.components.Action.primary.default.background,
          },
        },
      },
      primary: {
        'base': {
          backgroundColor: t.components.Action.primary.default.background,
          borderColor: t.components.Action.primary.default.background,
        },
        ':hover': {
          backgroundColor: t.components.Action.primary.hover.background,
          borderColor: t.components.Action.primary.hover.background,
        },
        ':active': {
          backgroundColor: t.components.Action.primary.active.background,
          borderColor: t.components.Action.primary.active.background,
        },
        ':disabled': {
          backgroundColor: t.components.Action.primary.disabled.background,
          borderColor: t.components.Action.primary.disabled.background,
        },

        'web': {
          ':focus-visible': {
            outlineColor: t.components.Action.primary.default.background,
          },
        },
      },
      secondary: {
        'base': {
          borderColor: t.components.Action.secondary.default.border,
          backgroundColor: t.components.Action.secondary.default.background,
        },
        ':hover': {
          borderColor: t.components.Action.secondary.hover.border,
          backgroundColor: t.components.Action.secondary.hover.background,
        },
        ':active': {
          borderColor: t.components.Action.secondary.active.border,
          backgroundColor: t.components.Action.secondary.active.background,
        },
        ':disabled': {
          borderColor: t.components.Action.secondary.disabled.border,
          backgroundColor: t.components.Action.secondary.disabled.background,
        },
      },
      tertiary: {
        'base': {
          borderColor: t.components.Action.tertiary.default.border,
          backgroundColor: t.components.Action.tertiary.default.background,
        },
        ':hover': {
          borderColor: t.components.Action.tertiary.hover.border,
          backgroundColor: t.components.Action.tertiary.hover.background,
        },
        ':active': {
          borderColor: t.components.Action.tertiary.active.border,
          backgroundColor: t.components.Action.tertiary.active.background,
        },
        ':disabled': {
          borderColor: t.components.Action.tertiary.disabled.border,
          backgroundColor: t.components.Action.tertiary.disabled.background,
        },
      },
    }) as const
);
