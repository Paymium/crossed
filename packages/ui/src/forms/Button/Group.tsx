/**
 * Copyright (c) Paymium.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root of this projects source tree.
 */

import { forwardRef } from 'react';
import { ProviderGroup, ButtonGroupCollection } from './context';
import { Orientation, RovingFocusGroup } from '@crossed/primitive';
import { type XBoxProps, XBox } from '../../layout/XBox';
import { View } from 'react-native';
import React from 'react';

export type ButtonGroupProps = XBoxProps & { orientation?: Orientation };
export const ButtonGroup = forwardRef<View, ButtonGroupProps>(
  (props: ButtonGroupProps, ref) => (
    <ProviderGroup grouped orientation={props.orientation ?? 'horizontal'}>
      <RovingFocusGroup orientation={props.orientation ?? 'horizontal'}>
        <ButtonGroupCollection.Provider>
          <ButtonGroupCollection.Slot>
            <XBox {...(props as any)} ref={ref} />
          </ButtonGroupCollection.Slot>
        </ButtonGroupCollection.Provider>
      </RovingFocusGroup>
    </ProviderGroup>
  )
);
