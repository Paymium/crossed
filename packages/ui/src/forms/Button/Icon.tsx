/**
 * Copyright (c) Paymium.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root of this projects source tree.
 */

import { composeStyles, CrossedMethods } from '@crossed/styled';
import {
  cloneElement,
  isValidElement,
  PropsWithChildren,
  useContext,
  useMemo,
} from 'react';
import { buttonContext } from './context';
import {
  textDisabledStyles,
  textErrorDisabledStyles,
  textErrorStyles,
  textStyles,
} from './styles';

export const ButtonIcon = ({
  children,
  style,
}: PropsWithChildren<{ style?: CrossedMethods<any> }>) => {
  const { variant, error, state, disabled } = useContext(buttonContext);

  const color = useMemo(() => {
    return composeStyles(
      textStyles.default,
      variant && textStyles[variant],
      disabled && textStyles.disabled,
      disabled && variant && textDisabledStyles[variant],
      variant && error && textErrorStyles[variant],
      disabled && variant && error && textErrorDisabledStyles[variant],
      style
    ).style(state).style.color;
  }, [variant, error, disabled, state]);

  return isValidElement(children)
    ? cloneElement(children, { color } as any)
    : children;
};
ButtonIcon.displayName = 'Button.Icon';
