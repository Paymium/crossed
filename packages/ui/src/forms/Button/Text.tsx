/**
 * Copyright (c) Paymium.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root of this projects source tree.
 */

import { forwardRef, useContext, useEffect } from 'react';
import { Text, type TextProps } from '../../typography/Text';
import { buttonContext } from './context';
import { composeStyles } from '@crossed/styled';
import {
  textDisabledStyles,
  textErrorDisabledStyles,
  textErrorStyles,
  textStyles,
} from './styles';
import React from 'react';
type ButtonTextProps = TextProps;

const ButtonText = forwardRef<any, ButtonTextProps>(
  (props: ButtonTextProps, ref) => {
    const { variant, error, state, disabled, setTextId, textId } =
      useContext(buttonContext);

    useEffect(() => {
      if (props.id && textId !== props.id) {
        setTextId(props.id);
      }
    }, [props.id, textId, setTextId]);

    return (
      <Text
        weight="lg"
        {...props}
        // disabled={disabled}
        {...state}
        id={textId}
        style={composeStyles(
          textStyles.default,
          variant && textStyles[variant],
          disabled && textStyles.disabled,
          disabled && variant && textDisabledStyles[variant],
          variant && error && textErrorStyles[variant],
          disabled && variant && error && textErrorDisabledStyles[variant],
          props.style
        )}
        ref={ref}
      />
    );
  }
);
ButtonText.displayName = 'Button.Text';

export { ButtonText, type ButtonTextProps };
