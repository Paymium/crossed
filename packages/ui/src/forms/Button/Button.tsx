/**
 * Copyright (c) Paymium.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root of this projects source tree.
 */

'use client';

import { composeStyles, withReactive } from '@crossed/styled';
import { ActivityIndicator, Pressable, View } from 'react-native';
import { forwardRef, useId, useState } from 'react';
import { buttonErrorStyles, buttonSizeStyles, buttonStyles } from './styles';
import { ButtonProps } from './types';
import { ButtonIcon } from './Icon';
import { buttonContext } from './context';

export const Button = withReactive(
  forwardRef<View, ButtonProps>(
    (
      {
        variant = 'primary',
        error = false,
        disabled,
        loading,
        size = true,
        children,
        ...props
      }: ButtonProps,
      ref
    ) => {
      const renderLoading = loading ? (
        <ButtonIcon>
          <ActivityIndicator />
        </ButtonIcon>
      ) : null;
      const id = useId();
      const [textId, setTextId] = useState(id);

      return (
        <Pressable
          aria-disabled={Boolean(disabled ?? false)}
          aria-labelledby={textId}
          role="button"
          tabIndex={disabled ? -1 : 0}
          disabled={disabled || loading}
          ref={ref}
          {...props}
          style={(e: any) =>
            composeStyles(
              buttonStyles.root,
              buttonSizeStyles[size.toString()],
              variant && buttonStyles[variant],
              (disabled || loading) &&
                variant &&
                buttonStyles[variant][':disabled'] && {
                  base: buttonStyles[variant][':disabled'],
                },
              error && buttonErrorStyles.error,
              variant && error && buttonErrorStyles[variant],
              (disabled || loading) &&
                variant &&
                error &&
                buttonErrorStyles[variant][':disabled'] && {
                  base: buttonErrorStyles[variant][':disabled'],
                },
              props.style
            ).rnw({
              hover: e.hovered,
              active: e.pressed,
              disabled: disabled || loading,
            }).style
          }
        >
          {(e: any) => {
            return (
              <buttonContext.Provider
                value={{
                  variant,
                  error,
                  state: { active: e.pressed, hover: e.hovered },
                  disabled: disabled || loading,
                  textId,
                  setTextId,
                }}
              >
                {renderLoading}
                {typeof children === 'function' ? children(e) : children}
              </buttonContext.Provider>
            );
          }}
        </Pressable>
      );
    }
  )
);
Button.displayName = 'Button';
