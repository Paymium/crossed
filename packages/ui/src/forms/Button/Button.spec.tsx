import * as React from 'react';
import { fireEvent, render, screen } from '@testing-library/react';

import { composeStories, setProjectAnnotations } from '@storybook/react';
import { test, expect } from '@jest/globals';
import { axe, toHaveNoViolations } from 'jest-axe';

import * as stories from './Button.stories'; // ðŸ‘ˆ Our stories imported here.

expect.extend(toHaveNoViolations);

const { Click } = composeStories(stories);

test('Checks if the form is valid', async () => {
  // Renders the composed story

  render(<Click />)

  const buttonElement = screen.getByRole('button', {
    name: 'Button',
  });

  expect(buttonElement).toMatchSnapshot("simple")

  fireEvent.pointerEnter(buttonElement);

  expect(buttonElement).toMatchSnapshot("hover")

  fireEvent.pointerDown(buttonElement);
  expect(buttonElement).toMatchSnapshot("active")


  fireEvent.pointerUp(buttonElement);
  expect(buttonElement).toMatchSnapshot("focus")

  fireEvent.pointerLeave(buttonElement);
  expect(buttonElement).toMatchSnapshot("focus")
});
