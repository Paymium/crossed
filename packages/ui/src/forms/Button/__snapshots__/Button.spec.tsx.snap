// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Checks if the form is valid: active 1`] = `
<button
  aria-labelledby=":r0:"
  class="css-view-175oi2r r-cursor-1loqt21 r-touchAction-1otgn73"
  id=":r0:"
  role="button"
  style="display: flex; border-top-left-radius: 7px; border-top-right-radius: 7px; border-bottom-right-radius: 7px; border-bottom-left-radius: 7px; border-top-width: 2px; border-right-width: 2px; border-bottom-width: 2px; border-left-width: 2px; border-top-style: solid; border-right-style: solid; border-bottom-style: solid; border-left-style: solid; flex-direction: row; align-items: center; justify-content: center; gap: 8px; height: 44px;"
  tabindex="0"
  type="button"
>
  <div
    class="css-text-146c3p1"
    dir="auto"
    id=":r0:"
    style="font-size: 16px; line-height: 20px; font-weight: bold; font-family: var(--font-family);"
  >
    Button
  </div>
</button>
`;

exports[`Checks if the form is valid: focus 1`] = `
<button
  aria-labelledby=":r0:"
  class="css-view-175oi2r r-cursor-1loqt21 r-touchAction-1otgn73"
  id=":r0:"
  role="button"
  style="display: flex; border-top-left-radius: 7px; border-top-right-radius: 7px; border-bottom-right-radius: 7px; border-bottom-left-radius: 7px; border-top-width: 2px; border-right-width: 2px; border-bottom-width: 2px; border-left-width: 2px; border-top-style: solid; border-right-style: solid; border-bottom-style: solid; border-left-style: solid; flex-direction: row; align-items: center; justify-content: center; gap: 8px; height: 44px;"
  tabindex="0"
  type="button"
>
  <div
    class="css-text-146c3p1"
    dir="auto"
    id=":r0:"
    style="font-size: 16px; line-height: 20px; font-weight: bold; font-family: var(--font-family);"
  >
    Button
  </div>
</button>
`;

exports[`Checks if the form is valid: focus 2`] = `
<button
  aria-labelledby=":r0:"
  class="css-view-175oi2r r-cursor-1loqt21 r-touchAction-1otgn73"
  id=":r0:"
  role="button"
  style="display: flex; border-top-left-radius: 7px; border-top-right-radius: 7px; border-bottom-right-radius: 7px; border-bottom-left-radius: 7px; border-top-width: 2px; border-right-width: 2px; border-bottom-width: 2px; border-left-width: 2px; border-top-style: solid; border-right-style: solid; border-bottom-style: solid; border-left-style: solid; flex-direction: row; align-items: center; justify-content: center; gap: 8px; height: 44px;"
  tabindex="0"
  type="button"
>
  <div
    class="css-text-146c3p1"
    dir="auto"
    id=":r0:"
    style="font-size: 16px; line-height: 20px; font-weight: bold; font-family: var(--font-family);"
  >
    Button
  </div>
</button>
`;

exports[`Checks if the form is valid: hover 1`] = `
<button
  aria-labelledby=":r0:"
  class="css-view-175oi2r r-cursor-1loqt21 r-touchAction-1otgn73"
  id=":r0:"
  role="button"
  style="display: flex; border-top-left-radius: 7px; border-top-right-radius: 7px; border-bottom-right-radius: 7px; border-bottom-left-radius: 7px; border-top-width: 2px; border-right-width: 2px; border-bottom-width: 2px; border-left-width: 2px; border-top-style: solid; border-right-style: solid; border-bottom-style: solid; border-left-style: solid; flex-direction: row; align-items: center; justify-content: center; gap: 8px; height: 44px;"
  tabindex="0"
  type="button"
>
  <div
    class="css-text-146c3p1"
    dir="auto"
    id=":r0:"
    style="font-size: 16px; line-height: 20px; font-weight: bold; font-family: var(--font-family);"
  >
    Button
  </div>
</button>
`;

exports[`Checks if the form is valid: simple 1`] = `
<button
  aria-labelledby=":r0:"
  class="css-view-175oi2r r-cursor-1loqt21 r-touchAction-1otgn73"
  id=":r0:"
  role="button"
  style="display: flex; border-top-left-radius: 7px; border-top-right-radius: 7px; border-bottom-right-radius: 7px; border-bottom-left-radius: 7px; border-top-width: 2px; border-right-width: 2px; border-bottom-width: 2px; border-left-width: 2px; border-top-style: solid; border-right-style: solid; border-bottom-style: solid; border-left-style: solid; flex-direction: row; align-items: center; justify-content: center; gap: 8px; height: 44px;"
  tabindex="0"
  type="button"
>
  <div
    class="css-text-146c3p1"
    dir="auto"
    id=":r0:"
    style="font-size: 16px; line-height: 20px; font-weight: bold; font-family: var(--font-family);"
  >
    Button
  </div>
</button>
`;
